README.md
# multi-progress-bars
![npm shield](https://img.shields.io/npm/v/multi-progress-bars) ![node shield](https://img.shields.io/npm/dm/multi-progress-bars)

A node utility written in Typescript for displaying multiple progress bars, and preserving logged output. The progress bars can be anchored to the top or bottom of the terminal, with border options and indefinite progress display. A Promise will resolve when all tasks are done. This utility is enabled by a simple virtual console that intercepts the normal console. See the below demo .gifs.

___

Anchoring to the top. Notice how the logs are preserved:

![MultiProgressBar top demo](./assets/mpb_top.gif)

---

Anchoring to the bottom:

![MultiProgressBar bottom demo](./assets/mpb_bottom.gif)

---

Showing off borders customizability and overflow management:

![MultiProgressBar extended demo](./assets/mpb_extended.gif)


## Install
`npm install multi-progress-bars`<br>
or<br>
`yarn add multi-progress-bars`

## Quickstart
```
import { MultiProgressBars } from 'multi-progress-bars';
import * as chalk from 'chalk';

// Initialize mpb
const mpb = new MultiProgressBars({
    initMessage: ' $ Example Fullstack Build ',
    anchor: 'top',
    persist: true,
    border: true,
});

// Add tasks
mpb.addTask('Webpack Backend', { type: 'percentage', barColorFn: chalk.yellow });
mpb.addTask('Watcher', { type: 'indefinite', message: 'Watching for changes...' });

// Update tasks
mpb.updateTask('Webpack Backend', { percentage: 0.2 });
mpb.incrementTask('Webpack Backend', { percentage: 0.1 });

// Console logging is overridden and tracked with an interval buffer
console.log('Warning!');

setTimeout(() => { mpb.done('Webpack Backend', { message: 'Build finished.' })}, 5000);
setTimeout(() => { mpb.done('Watcher', { message: 'Build finished.' })}, 1000);

// Wait for all tasks to finish
await mpb.promise;
console.log('Finish');

```
See the [Documentation](#documentation) for more information.


## Version Changes

<br>

> ## :warning: v4 -> v5 has **breaking changes**.<br>
>* Library now defaults to ESM. Dependencies on chalk and related libraries are now updated to the latest, which default to ESM.
>* For building, node needs to be >14.18 because of Rollup 3

<br>

> ## :warning: v3 -> v4 has **breaking changes**.<br>
> **User-Facing Changes**:
>* `barColorFn` is now `barTransformFn`, and a new property `nameTransformFn` has been added to transform task names appearance.
>* Number of progress bars displayed will be truncated to terminal height (accounting for borders)
>* Upon exiting, will dump the full untruncated list of progress bars if the above was true.
>* API added for removing tasks, with option to shift up the tasks or leave a blank line.
>* API added for getting the index of a specific task, or getting the name of a task given the index.
>* Extra border options in constructor as well as API to set or remove headers and footers.
>* Bottom-anchored progress bars now has bottom border if `border` is set to `true` in the constructor.

<br>

> :warning: v2 -> v3 technically has **breaking changes** in behavior nuance, but should be backwards-compatible. The changes are the addition of anchor position, border, and persist. If you pass in `true` for persist, you should call `mpb.close()` when you are done, or on `SIGINT` for example. Multi-line init messages are now clamped to the first line.

<br>

## <span name="documentation">Documentation</span>

* [Usage guide](docs/Usage.md)
* [API guide](docs/API.md)
* [Changelog](CHANGELOG.md)
* [Examples](docs/Examples.md)

## Hilbert Spinner

Included in this library is a (imo) cool Hilbert Curve / Space-Filling Curve spinner. It uses the Braille dots to do a little snake that crawls throughout the bar. There is probably a more efficient way to code the crawling, as this uses sort of an 'implicit' approach, instead of an 'explicit', which might require less calculation.

## License

[MIT](LICENSE)

## Like this utility? Support my hobby!

[PayPal donation link](https://paypal.me/seanchenpiano?locale.x=en_US)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [commit-and-tag-version](https://github.com/absolute-version/commit-and-tag-version) for commit guidelines.

## [5.0.3](https://github.com/kamiyo/multi-progress-bars/compare/v5.0.2...v5.0.3) (2023-01-24)

### [5.0.2](https://github.com/kamiyo/multi-progress-bars/compare/v5.0.1...v5.0.2) (2022-12-22)


### Bug Fixes

* **core:** Incorrect removal of borders. Also example was introducing bug because of delay/promise. ([2de9c58](https://github.com/kamiyo/multi-progress-bars/commit/2de9c58b5dfc06b08f0a1bb68407d347bbf3926a))

### [5.0.1](https://github.com/kamiyo/multi-progress-bars/compare/v5.0.0...v5.0.1) (2022-12-22)


### Bug Fixes

* **readme:** Updated README with breaking change notification. ([487a7c9](https://github.com/kamiyo/multi-progress-bars/commit/487a7c908e2bbe87c15d2a77c042bc573062e927))

## [5.0.0](https://github.com/kamiyo/multi-progress-bars/compare/v4.3.4...v5.0.0) (2022-12-22)


### ⚠ BREAKING CHANGES

* **core:** Update to using ESM. Update Rollup to v3, requiring rollup config as mjs. Update chalk, string-width, and strip-ansi to latest, which use ESM. Example is now ESM.

### Features

* **core:** Update to using ESM. Update Rollup to v3, requiring rollup config as mjs. Update chalk, string-width, and strip-ansi to latest, which use ESM. Example is now ESM. ([591e001](https://github.com/kamiyo/multi-progress-bars/commit/591e001cdb26f400d2fc2e93a475871c6dfb5f8f))

### [4.3.4](https://github.com/kamiyo/multi-progress-bars/compare/v4.3.3...v4.3.4) (2022-12-22)


### Bug Fixes

* **core:** Complete updating yarn and sdks. ([9eba2dc](https://github.com/kamiyo/multi-progress-bars/commit/9eba2dc1f7d80e076577f23b3bf25cae43a15b2d))

### [4.3.3](https://github.com/kamiyo/multi-progress-bars/compare/v4.3.2...v4.3.3) (2022-12-22)

### [4.3.2](https://github.com/kamiyo/multi-progress-bars/compare/v4.3.1...v4.3.2) (2022-12-22)


### Bug Fixes

* **core:** Update dependencies to minor latest minor version as well as necessary security updates. ([9e7ddbf](https://github.com/kamiyo/multi-progress-bars/commit/9e7ddbf264bdf009a202a098c518f1ea3b75a984))

### [4.3.1](https://github.com/kamiyo/multi-progress-bars/compare/v4.3.0...v4.3.1) (2022-07-12)


### Bug Fixes

* **deps:** deduped deps ([5b86445](https://github.com/kamiyo/multi-progress-bars/commit/5b86445713a1f2cb6dd7d2c75d7416684ea26d38))

## [4.3.0](https://github.com/kamiyo/multi-progress-bars/compare/v4.2.3...v4.3.0) (2022-07-12)


### Bug Fixes

* **deps:** updated dependencies and nested dependencies for security ([6643a81](https://github.com/kamiyo/multi-progress-bars/commit/6643a811dd2cc2eed9b33120b922189f5f91efd0))

### [4.2.3](https://github.com/kamiyo/multi-progress-bars/compare/v4.2.2...v4.2.3) (2022-01-13)


### Bug Fixes

* **core:** add carriage return to line split matching ([9c23705](https://github.com/kamiyo/multi-progress-bars/commit/9c237055e58786389340ac1a12da90c0bdb9a3c7))

### [4.2.2](https://github.com/kamiyo/multi-progress-bars/compare/v4.2.2-0...v4.2.2) (2022-01-13)

### [4.2.2-0](https://github.com/kamiyo/multi-progress-bars/compare/v4.2.1...v4.2.2-0) (2022-01-09)


### Bug Fixes

* **core:** Maybe fix ESM? ([cfab46e](https://github.com/kamiyo/multi-progress-bars/commit/cfab46e2033da2c64fa25b9c2c47c9336ea61f00))

### [4.2.1](https://github.com/kamiyo/multi-progress-bars/compare/v4.2.0...v4.2.1) (2022-01-09)


### Bug Fixes

* **core:** Fix empty string crashing border ([871633b](https://github.com/kamiyo/multi-progress-bars/commit/871633b265d6efddec0e8428a8f4fc78a36a5d34))
* **core:** Fix newlines and tabs breaking border message. Now splits at those characters and only displays the first item of the returned array. ([1e09301](https://github.com/kamiyo/multi-progress-bars/commit/1e093013483194654731d11c961b69a9f74f73f7))

## [4.2.0](https://github.com/kamiyo/multi-progress-bars/compare/v4.1.0...v4.2.0) (2021-11-09)


### Features

* **docs:** Added missing header/footer API methods. ([9f8f3e2](https://github.com/kamiyo/multi-progress-bars/commit/9f8f3e215ee1af0ee484db135aba459a84bb8c66))

## [4.1.0](https://github.com/kamiyo/multi-progress-bars/compare/v4.0.1...v4.1.0) (2021-11-09)


### Features

* **dev:** Update yarn to v2.4.2 ([f723ace](https://github.com/kamiyo/multi-progress-bars/commit/f723ace00659de995acc150fb57689cda2f0a223))


### Bug Fixes

* **core:** removeBottomBorder had outdated progressHeight calculations ([6b1332f](https://github.com/kamiyo/multi-progress-bars/commit/6b1332fd3c2213ef325498ca7321dcc23de53148))
* **package:** Change esm extension to mjs for automatic module detection ([6107343](https://github.com/kamiyo/multi-progress-bars/commit/61073430682c3f0622b830a4694242f0f11c19c0))

### [4.0.1](https://github.com/kamiyo/multi-progress-bars/compare/v4.0.0...v4.0.1) (2021-06-30)


### Bug Fixes

* **core:** resizing logic ([2b09ba5](https://github.com/kamiyo/multi-progress-bars/commit/2b09ba561e705a00c38542a87c04323628541c88))
* **core:** update typescript and yarn version ([9fc6147](https://github.com/kamiyo/multi-progress-bars/commit/9fc6147596f78776e2fff71cbfe44be6736579a0))
* **core:** upgrade yarn stuff ([79829cf](https://github.com/kamiyo/multi-progress-bars/commit/79829cf8bc0daee5bed1e12450d1af13880066ff))

## [4.0.0](https://github.com/kamiyo/multi-progress-bars/compare/v4.0.0-alpha.0...v4.0.0) (2021-06-23)

### Bug Fixes

* **dep:** Update node requirements ([92d95d5](https://github.com/kamiyo/multi-progress-bars/commit/92d95d5a329b354fdd4965ed6939102b78c22202))
* **dep:** Pin string-width and strip-ansi version to before they moved to es-modules

* **core:** promise resetting was broken
* **core:** bottom-anchored progress bars now has bottom border as well.

### Features

* **core:** barColorFn renamed to barTransformFn
* **core:** added nameTransformFn property
* **core:** added removeTask
* **core:** added extended border options
* **core:** crawler improvements

examples/example.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const multi_progress_bars_1 = require("multi-progress-bars");
const chalk_1 = __importDefault(require("chalk"));
let timerId, timeoutId;
const testBasic = (anchor) => __awaiter(void 0, void 0, void 0, function* () {
    const task1 = 'Build App';
    const task2 = 'Build Server';
    const taskInf = 'Watcher';
    const mpb = new multi_progress_bars_1.MultiProgressBars({
        initMessage: ' $ Example Fullstack Build ',
        anchor,
        persist: true,
        progressWidth: 40,
        numCrawlers: 7,
        border: true,
    });
    process.on('SIGINT', () => {
        clearInterval(timerId);
        clearTimeout(timeoutId);
        process.exit();
    });
    mpb.addTask(task1, { type: 'percentage', barTransformFn: chalk_1.default.yellow, nameTransformFn: chalk_1.default.green });
    mpb.addTask(task2, { type: 'percentage', barTransformFn: chalk_1.default.blue, nameTransformFn: chalk_1.default.yellow });
    mpb.addTask(taskInf, { type: 'indefinite', barTransformFn: chalk_1.default.magenta, message: 'Linting started.' });
    let timerCount = 0;
    timerId = setInterval(() => {
        if (timerCount % 2 === 0) {
            mpb.incrementTask(task1, { percentage: 0.05 });
            console.log("[Frontend] log: ", timerCount);
        }
        if (timerCount % 8 === 0) {
            console.log(chalk_1.default.yellow("[Backend] warning: ", timerCount));
        }
        if (timerCount % 16 === 0) {
            console.log("[Frontend] info:\n  app.js:%d:%d\n  Missing stuff.\n  Fixit!", timerCount % 16, timerCount % 23);
        }
        mpb.incrementTask(task2, { percentage: 0.04 });
        timerCount++;
    }, 200);
    timeoutId = setTimeout(() => {
        mpb.done(taskInf, { message: 'Linting finished. Output in eslint-output.txt' });
        timeoutId = setTimeout(() => {
            mpb.updateTask(taskInf, { message: 'Linting started.' });
            timeoutId = setTimeout(() => {
                mpb.done(taskInf);
            }, 3000);
        }, 2000);
    }, 5000);
    yield mpb.promise;
    clearInterval(timerId);
    mpb.close();
    console.log("Build exited.");
});
const testOverflow = (anchor) => __awaiter(void 0, void 0, void 0, function* () {
    const mpb = new multi_progress_bars_1.MultiProgressBars({
        initMessage: ' $ Overflow Test ',
        anchor,
        persist: true,
        border: '\u2550',
    });
    let count = 0;
    let done = 0;
    const addTaskTimerId = setInterval(() => {
        const taskName = 'Task ' + count;
        mpb.addTask(taskName, { type: 'percentage', percentage: 0 });
        mpb.setFooter({
            pattern: '\u2550',
            message: ` ${done}/${count} | ${(100 * done / count).toFixed(0)}% `,
            right: 4
        });
        const innerTimerId = setInterval(() => {
            mpb.incrementTask(taskName, { percentage: 0.05 });
            if (mpb.isDone(taskName)) {
                // mpb.removeTask(taskName);
                done++;
                mpb.setFooter({
                    message: ` ${done}/${count} | ${(100 * done / count).toFixed(0)}% `,
                });
                clearInterval(innerTimerId);
            }
        }, 100);
        count++;
        if (count == 40) {
            clearInterval(addTaskTimerId);
        }
        if (count % 5 === 0) {
            console.log('Console log: %d', count);
        }
    }, 200);
    console.log('Initial console logs:\n1\n2\n3\n4\n5');
    yield mpb.promise;
    mpb.close();
});
const testRemove = (anchor) => __awaiter(void 0, void 0, void 0, function* () {
    const mpb = new multi_progress_bars_1.MultiProgressBars({
        anchor,
        persist: true,
        header: { pattern: '*', message: ' Removal Test ' },
        footer: { pattern: '.' },
    });
    let count = 0;
    const addTaskTimerId = setInterval(() => {
        const taskName = 'Task ' + count;
        mpb.addTask(taskName, { type: 'percentage', percentage: 0 });
        const innerTimerId = setInterval(() => {
            mpb.incrementTask(taskName, { percentage: 0.05 });
            if (mpb.isDone(taskName)) {
                setTimeout(() => {
                    mpb.removeTask(taskName);
                }, 1000);
                clearInterval(innerTimerId);
            }
        }, 100);
        count++;
        if (count == 40) {
            clearInterval(addTaskTimerId);
        }
        if (count % 5 === 0) {
            console.log("console log:\n%d", count);
        }
    }, 200);
    console.log('Initial console logs:\n1\n2\n3\n4\n5');
    yield mpb.promise;
    mpb.close();
});
const testBorders = (anchor) => __awaiter(void 0, void 0, void 0, function* () {
    const mpb = new multi_progress_bars_1.MultiProgressBars({
        anchor,
        header: { message: ' Borders Test ', pattern: '*' },
        footer: { pattern: '.' },
        persist: true,
    });
    let count = 0;
    const addTaskTimerId = setInterval(() => {
        const taskName = 'Task ' + count;
        mpb.addTask(taskName, { type: 'percentage', percentage: 0 });
        const innerTimerId = setInterval(() => {
            mpb.incrementTask(taskName, { percentage: 0.05 });
            if (mpb.isDone(taskName)) {
                clearInterval(innerTimerId);
            }
        }, 100);
        count++;
        if (count == 20) {
            clearInterval(addTaskTimerId);
        }
        if (count % 3 === 0) {
            console.log("console log:\n%d", count);
            if (count % 2 === 1) {
                mpb.setHeader({ pattern: ' ' });
                mpb.setFooter({ pattern: ' ' });
            }
            else {
                mpb.setHeader({ pattern: '*' });
                mpb.setFooter({ pattern: '.' });
            }
        }
    }, 200);
});
(() => __awaiter(void 0, void 0, void 0, function* () {
    let anchor = 'top';
    if (process.argv.includes('--bottom')) {
        anchor = 'bottom';
    }
    if (process.argv.includes('--basic') || process.argv.includes('-b')
        || process.argv.includes('--all') || process.argv.includes('-a')
        || process.argv.length === 2) {
        yield testBasic(anchor);
    }
    if (process.argv.includes('--overflow') || process.argv.includes('-o')
        || process.argv.includes('--all') || process.argv.includes('-a')
        || process.argv.length === 2) {
        yield testOverflow(anchor);
    }
    if (process.argv.includes('--remove') || process.argv.includes('-r')
        || process.argv.includes('--all') || process.argv.includes('-a')
        || process.argv.length === 2) {
        yield testRemove(anchor);
    }
    if (process.argv.includes('--borders') || process.argv.includes('-e')
        || process.argv.includes('--all') || process.argv.includes('-a')
        || process.argv.length === 2) {
        yield testBorders(anchor);
    }
}))();
